File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.cpp
#
# @lc app=leetcode id=3768 lang=cpp
#
# [3768] Minimum Inversion Count in Subarrays of Fixed Length
#
# @lc code=start
class Solution {
public:
    long long minInversionCount(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> sorted_nums = nums;
        sort(sorted_nums.begin(), sorted_nums.end());
        unordered_map<int, int> rank_map;
        for (int i = 0; i < n; ++i) {
            rank_map[sorted_nums[i]] = i + 1; // Rank starts from 1
        }
        
        vector<int> bit(n + 1, 0);
        auto update = [&](int index, int value) {
            while (index <= n) {
                bit[index] += value;
                index += index & -index;
            }
        };
        
        auto query = [&](int index) {
            int sum = 0;
            while (index > 0) {
                sum += bit[index];
                index -= index & -index;
            }
            return sum;
        };

        long long current_inversions = 0;

        // Initial window setup
        for (int i = 0; i < k; ++i) {
            int rank = rank_map[nums[i]];
            current_inversions += query(n) - query(rank);
            update(rank, 1);
        }

        long long min_inversions = current_inversions;

        // Sliding window
        for (int i = k; i < n; ++i) {
            int rank_outgoing = rank_map[nums[i-k]];
            update(rank_outgoing, -1);
            current_inversions -= query(n) - query(rank_outgoing);

            int rank_incoming = rank_map[nums[i]];
            current_inversions += query(n) - query(rank_incoming);
            update(rank_incoming, 1);

            min_inversions = min(min_inversions, current_inversions);
        }

        return min_inversions;
    }
};
# @lc code=end

File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.go
//
// @lc app=leetcode id=3768 lang=golang
//
// [3768] Minimum Inversion Count in Subarrays of Fixed Length
//

// @lc code=start
func minInversionCount(nums []int, k int) int64 {
    n := len(nums)
    if k == 1 {
        return 0 // No inversions possible in single-element subarrays
    }
    
    // Define a Fenwick Tree structure
    type FenwickTree struct {
        tree []int64
        size int
    }
    
    // Initialize Fenwick Tree
    func (ft *FenwickTree) init(size int) {
        ft.tree = make([]int64, size+1)
        ft.size = size
    }
    
    // Update function for Fenwick Tree
    func (ft *FenwickTree) update(index int, delta int64) {
        for i := index; i <= ft.size; i += i & (-i) {
            ft.tree[i] += delta
        }
    }
    
    // Prefix sum query for Fenwick Tree
    func (ft *FenwickTree) query(index int) int64 {
        var sum int64 = 0
        for i := index; i > 0; i -= i & (-i) {
            sum += ft.tree[i]
        }
        return sum
    }
    
    // Coordinate compression map to ensure indices fit within tree size limits
    coordinateMap := make(map[int]int)
    sortedNums := make([]int, n)
    copy(sortedNums, nums)
    sort.Ints(sortedNums)
    idx := 1 // Start indices from 1 for Fenwick Tree compatibility
    for _, num := range sortedNums {
        if _, exists := coordinateMap[num]; !exists {
            coordinateMap[num] = idx
            idx++
        }
    }
    
    fenwickTree := FenwickTree{}
    fenwickTree.init(len(coordinateMap))
    															minInversions := int64(^uint64(0) >> 1) // Max integer value as initial min inversion count// Sliding Window Algorithm with Inversion Counting using a Fenwick Tree.for startIdx := 0; startIdx+k <= n; startIdx++ {if startIdx > 0 { prevIndexMapped := coordinateMap[nums[startIdx-1]] fenwickTree.update(prevIndexMapped, -1)} inversionsInWindow := int64(0)	for j := startIdx; j < startIdx+k; j++ {indexMapped := coordinateMap[nums[j]]inversionsInWindow += fenwickTree.query(fenwickTree.size) - fenwickTree.query(indexMapped)	fenwickTree.update(indexMapped, 1)}if inversionsInWindow < minInversions {minInversions = inversionsInWindow}}return minInversions}// @lc code=end